cpu/sam0_common/periph/dac.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2020 Beuth Hochschule für Technik Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_sam0_common
+ * @{
+ *
+ * @file
+ * @brief       Low-level DAC driver implementation
+ *
+ * @author      Benjamin Valentin <benpicco@beuth-hochschule.de>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "periph/dac.h"
+#include "periph/gpio.h"
+
+#include "byteorder.h"
+
+#ifndef DAC_CLOCK
+#define DAC_CLOCK   DAC_CLOCK_DEFAULT
+#endif
+
+#define DAC_VAL(in) (in >> (16 - DAC_RES_BITS))
```

Comment:
I don't get this.
Why playing with the bit position here ?
In the current state, DAC_DACCTRL_LEFTADJ is set to 0 and we don't use dithering so we should write data to bits [0..9] or  [0..11]. there is no need for this macro, just assign the value to the register directly.

Reply:
The problem is that the DAC API assumes a full 16-bit input value.
So setting a value > 1023 will make the DAC overflow back to zero - you can observe this with `tests/periph_dac`.

On other DACs (e.g. lpc2387) you write the full 16-bit value but this being a 10 bit DAC, the lower (least significant) 6 bytes are ignored.
Here the most significant bytes would be ignored, which leads to wrong output values.